### PR TITLE
Adding a mock State class that can be instantiated and queried while testing

### LINKED ISSUE #2098: Unit testing with State mock-up
### Description

The goal of this issue is to provide a simple way to test the code of a Taipy application.

This issue is related to: https://github.com/Avaiga/taipy/issues/1716

What can be tested:
- *on_change* callbacks
- *on_action* callbacks
- *on_edit*, *on_add* and other Taipy callbacks
- functions that have as a parameter, the *state*.

### Solution Proposed

A way to make this process easy is to create a mock-up of the *state* variable that can be used for unit testing.

This *state* should be retrieved from Taipy in some way. The goal would then to fill this *state* with variables and pass it to the callback function and check if the *state* returned is as expected. 

### Acceptance Criteria

- [ ] If applicable, a new demo code is provided to show the new feature in action.
- [ ] Integration tests exhibiting how the functionality works are added.
- [ ] Any new code is covered by a unit tested.
- [ ] Check code coverage is at least 90%.
- [ ] Related issue(s) in taipy-doc are created for documentation and Release Notes are updated.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
I had hoped it would not have any impact on the "real" code but it has. 
Any discussion is welcome before merge. 
resolves #2098

Here is how I think it could be used for tests (aiming at Taipy user developing and testing their app)

The user code to test
```py
def on_action(state: State):
    ...
    state.assign("a", 2)

```
- The user test code:
```py
def test_callback():
    ms = MockState(Gui(""), a = 1)
    on_action(ms)
    assert ms.a == 2

```

### CODE REVIEW

File: taipy/gui/state.py [hunk 1]
```diff
@@ -73,6 +74,87 @@ def change_variable(state):
     ```
     """
 
+    def __init__(self) -> None:
+        self._gui: "Gui"
+
+    @abstractmethod
```
Comment (dinhlongviolin1) [inline taipy/gui/state.py:80]: the State class should be abstract also

File: taipy/mock/gui/mock_state.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+import typing as t
+
+from ...gui import Gui, State
+from ...gui.utils import _MapDict
+
+
+class MockState(State):
+    __VARS = "vars"
```
Comment (dinhlongviolin1) [inline taipy/mock/gui/mock_state.py]: maybe a todo for docs here I am not sure :D

**Review by dinhlongviolin1 — COMMENTED** [review]
One small comment is the placement of the mock folder. I think it would be better to place it into the gui as it would simplify the packaging process.

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

File: taipy/gui/mock/mock_state.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    """
+
+    __VARS = "vars"
+
+    def __init__(self, gui: Gui, **kwargs) -> None:
+        super().__setattr__(MockState.__VARS, {k: _MapDict(v) if isinstance(v, dict) else v for k, v in kwargs.items()})
+        self._gui = gui
+        super().__init__()
+
+    def get_gui(self) -> "Gui":
```
Comment (FabienLelaquais) [inline taipy/gui/mock/mock_state.py]:
```suggestion
    def get_gui(self) -> Gui:
```

File: taipy/gui/mock/mock_state.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    def __setattr__(self, name: str, value: t.Any) -> None:
+        t.cast(dict, super().__getattribute__(MockState.__VARS))[name] = (
+            _MapDict(value) if isinstance(value, dict) else value
+        )
+
+    def __getitem__(self, key: str):
+        return self
+
+    def __enter__(self):
```
Comment (FabienLelaquais) [inline taipy/gui/mock/mock_state.py:55]: Doesn't this prevent MockState to be used as a context manager?
The tests seems to respond no...

**Review by FabienLelaquais — DISMISSED** [review]
🤔

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**